### PR TITLE
[GLUTEN-4084][CORE][VL] Verify supported file system

### PR DESCRIPTION
**Pull request by wForget**
## What changes were proposed in this pull request?

Verify supported file system

Fixes: #4084

## How was this patch tested?

added test.

After this pr:

unsupported file system:

image
image
image

supported file system:

image

### CODE REVIEW

File: gluten-core/src/main/scala/org/apache/spark/sql/hive/HiveTableScanExecTransformer.scala [hunk 1]
```diff
@@ -79,6 +80,10 @@ class HiveTableScanExecTransformer(
     Seq.empty
   }
 
+  override def getRootPathsInternal: Seq[Path] = {
+    relation.tableMeta.storage.locationUri.map(new Path(_)).toSeq
```
Comment (PHILO-HE) [inline gluten-core/src/main/scala/org/apache/spark/sql/hive/HiveTableScanExecTransformer.scala:84]: Thanks for your work!
Could you clarify the overhead required for validating root paths? Does each file have a root path?

Comment (wForget) [reply gluten-core/src/main/scala/org/apache/spark/sql/hive/HiveTableScanExecTransformer.scala:84]: I think the overhead of rootPath is smaller, at least it doesn't need to be listed. For nodes without rootPath we will pass conversion.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala [hunk 2]
```diff
@@ -53,9 +56,10 @@ object VeloxBackend {
   val BACKEND_NAME = "velox"
 }
 
-object BackendSettings extends BackendSettingsApi {
+object BackendSettings extends BackendSettingsApi with Logging {
 
   val SHUFFLE_SUPPORTED_CODEC = Set("lz4", "zstd")
+  val SUPPORTED_FILE_SYSTEM = Set("file", "hdfs", "s3a")
```
Comment (leesf) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:62]: oss can also be supported via s3a scheme.

Comment (PHILO-HE) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:62]: Maybe, we should include all schemes in the below link and name this set `SUPPORTED_SCHEMES`. @leesf, does it make sense?
https://github.com/facebookincubator/velox/blob/aaaa0791008267d8be0a5c141db6da261aba8892/velox/connectors/hive/storage_adapters/s3fs/S3Util.h#L34-L45

Comment (PHILO-HE) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:62]: If it is named `SUPPORTED_SCHEMES`, maybe each scheme name should be xxx://.

Comment (wForget) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:62]: how about we provide a blacklist configuration to exclude?

Comment (wForget) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:62]: all schemes:

```
Set(
 "file", // local
 "hdfs", // hdfs
 "s3", "s3a", "s3n", "oss", "oss", "cosn",  // s3
 "gs", // gcs
 "abfs", "abfss"  // abfs
) 
```

Comment (wForget) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:62]: > If it is named `SUPPORTED_SCHEMES`, maybe each scheme name should be xxx://.

Do we need that? The scheme obtained by `java.net.URI#getScheme` does not have the '://' suffix.

Comment (PHILO-HE) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:62]: Maybe, using a configurable blacklist is not necessary.

BTW, I note velox registers S3/HDFS/GCS/ABFS according to the corresponding build option. If the validation passes on Gluten side, but velox doesn't actually register the file system. Can the expected fallback still happen?
https://github.com/facebookincubator/velox/blob/aaaa0791008267d8be0a5c141db6da261aba8892/velox/connectors/hive/HiveConnector.cpp#L151-L159

Comment (PHILO-HE) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala]: > > If it is named `SUPPORTED_SCHEMES`, maybe each scheme name should be xxx://.
> 
> Do we need that? The scheme obtained by `java.net.URI#getScheme` does not have the '://' suffix.

OK to me if "://" is lacking.

Comment (wForget) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala]: > BTW, I note velox registers S3/HDFS/GCS/ABFS according to the corresponding build option. If the validation passes on Gluten side, but velox doesn't actually register the file system. Can the expected fallback still happen?

Good point. How about we add relevant comments in the code? It prompts to recompile velox for support.

Comment (PHILO-HE) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala]: OK. Please add some comments to clarify. Thanks!

Comment (wForget) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala]: > OK. Please add some comments to clarify. Thanks!

Added, please help to continue the review, thanks.

**Comment by PHILO-HE** [comment]
@rui-mo, do you have any comment?

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala (hunk 2 above)
Comment (leesf) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala]: @PHILO-HE agree that we should not break the capabilities velox already have. In addition, if users add new schemes to velox(which is compatible with s3a), we must also fix it in gluten, which would make it a little complex. But I am open with PR.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala [hunk 3]
```diff
@@ -53,18 +58,46 @@ object VeloxBackend {
...
+    // local
+    "file",
+    // hdfs
+    "hdfs",
+    // s3
+    "s3",
+    "s3a",
+    "s3n",
+    "oss",
+    "oss",
```
Comment (leesf) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:83]: duplicate items.

Comment (wForget) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala]: Fixed, thanks.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala (hunk 2 above)
Comment (PHILO-HE) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala]: >  if users add new schemes to velox(which is compatible with s3a), we must also fix it in gluten

For making file system support state always consistent with velox, one way is to do the validation on native side, i.e., calling velox [getFileSystem](https://github.com/facebookincubator/velox/blob/5144f141cfe2cc832d324085ed570457c22a9a77/velox/common/file/FileSystems.cpp#L52) to validate. @wForget, would you like to improve this PR?

Comment (wForget) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala]: Sure, I will try it.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala [hunk 4]
```diff
@@ -83,33 +116,43 @@ object BackendSettings extends BackendSettingsApi {
...
+        logWarning(
           s"Validation failed for ${this.getClass.toString}" +
             s" due to: data type $unsupportedDataType. in file schema. ")
-        // scalastyle:on println
       }
       unsupportedDataTypes.isEmpty
     }
 
+    // check if supported file system
+    val unsupportedPath = rootPaths.find(!isSupportedFileSystem(_))
```
Comment (rui-mo) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:127]: Do we need to verify all the root paths? Is it a usual case for files belonging to different systems?

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala [hunk 5]
```diff
@@ -307,4 +350,6 @@ object BackendSettings extends BackendSettingsApi {
   override def staticPartitionWriteOnly(): Boolean = true
 
   override def allowDecimalArithmetic: Boolean = SQLConf.get.decimalOperationsAllowPrecisionLoss
+
+  override def requiredRootPaths(): Boolean = true
```
Comment (rui-mo) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:354]: Do we need this variable, and in what case the root paths are not required?

**Review by rui-mo — COMMENTED** [review]
Thanks for working on this. Just two quesitons.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala (hunk 4 above)
Comment (wForget) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:127]: > Do we need to verify all the root paths? Is it a usual case for files belonging to different systems?

Maybe first is enough. I think the rootPaths of one datasource are mostly the same filesystem.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala (hunk 5 above)
Comment (wForget) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:354]: > Do we need this variable, and in what case the root paths are not required?

I'm not familiar with clickhouse backend so I kept this and enabled it only for velox backend.

Comment (rui-mo) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:354]: @zzcclp Could you help check if it's needed for CH backend? Thanks.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala (hunk 4 above)
Comment (rui-mo) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:127]: @PHILO-HE What do you think?

Comment (PHILO-HE) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:127]: Maybe, multiple kinds of file system are not commonly used in scan. We may just validate the first one. If doing so, let's leave some comment to clarify.

File: backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHBackend.scala [hunk 6]
```diff
@@ -129,7 +131,8 @@ object CHBackendSettings extends BackendSettingsApi with Logging {
       format: ReadFileFormat,
       fields: Array[StructField],
       partTable: Boolean,
-      paths: Seq[String]): Boolean = {
+      paths: Seq[String],
+      rootPaths: Seq[Path]): Boolean = {
```
Comment (zhztheplayer) [inline backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHBackend.scala:135]: Minor: Do we have to have `paths` and  `rootPaths` together in the parameter list? Would it be reasonable to use the latter to extract the former?

Comment (wForget) [reply backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHBackend.scala:135]: `paths` has a large initialization cost and is empty by default in velox backend.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala [hunk 7]
```diff
@@ -53,18 +58,46 @@ object VeloxBackend {
...
 
+  // scalastyle:off line.size.limit
+  /**
+   * Supported file schemes for Velox backend.
+   *
+   * Note: Need to turn on the required file systems when compiling velox to obtain support. Refer:
+   * https://github.com/facebookincubator/velox/blob/fa6de06c826084ce5d938c71b7450c59204c864d/CMakeLists.txt#L93-L96
+   */
+  // scalastyle:on line.size.limit
+  val SUPPORTED_FILE_SCHEMES = Set(
```
[the remark this answers is not in the document]
Comment (wForget) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:73]: Maybe you missed the discussion in https://github.com/oap-project/gluten/pull/4086#discussion_r1430817063. We will check if filesystem is supported via native method. What do you think?

[the remark this answers is not in the document]
Comment (zhli1142015) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala:73]: I see.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala (hunk 4 above)
Comment (wForget) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala]: Fixed, thanks

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala (hunk 2 above)
Comment (wForget) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/VeloxBackend.scala]: > > if users add new schemes to velox(which is compatible with s3a), we must also fix it in gluten
> 
> For making file system support state always consistent with velox, one way is to do the validation on native side, i.e., calling velox [getFileSystem](https://github.com/facebookincubator/velox/blob/5144f141cfe2cc832d324085ed570457c22a9a77/velox/common/file/FileSystems.cpp#L52) to validate. @wForget, would you like to improve this PR?

@PHILO-HE  I added a native method to check paths, could you please help me take a look?

**Comment by wForget** [comment]
@PHILO-HE @rui-mo could you please take a look again if you have time?

**Comment by PHILO-HE** [comment]
Thanks so much for the iterations! Can we do the validation in `SubstraitToVeloxPlanValidator`? I think it would be better.

**Comment by wForget** [comment]
> Thanks so much for the iterations! Can we do the validation in `SubstraitToVeloxPlanValidator`? I think it would be better.

Thanks for your guidance I will try that.

**Comment by zhli1142015** [comment]
Hello @wForget,
I saw this PR is closed, will you continue this PR?
Thanks.

**Comment by wForget** [comment]
> Hello @wForget, I saw this PR is closed, will you continue this work? Thanks.

I disabled columnar filescan by adding `spark.gluten.sql.columnar.filescan=false;` config, so this PR is unnecessary for me, if this is useful I can continue to work on it. cc @PHILO-HE WDYT?

**Comment by zhli1142015** [comment]
We need this PR, when reading from for not supported file systems, falling back is always better than failing.

**Comment by PHILO-HE** [comment]
> > Hello @wForget, I saw this PR is closed, will you continue this work? Thanks.
> 
> I disabled columnar filescan by adding `spark.gluten.sql.columnar.filescan=false;` config, so this PR is unnecessary for me, if this is useful I can continue to work on it. cc @PHILO-HE WDYT?

@wForget, do you remember the landing gap of this pr? Thanks!

**Comment by wForget** [comment]
> @wForget, do you remember the landing gap of this pr? Thanks!

I remember the current status should be waiting for review, but I didn't move forward with it because I found `spark.gluten.sql.columnar.filescan=false` configuration. Do we need to reopen this PR?

**Comment by PHILO-HE** [comment]
> > @wForget, do you remember the landing gap of this pr? Thanks!
> 
> I remember the current status should be waiting for review, but I didn't move forward with it because I found `spark.gluten.sql.columnar.filescan=false` configuration. Do we need to reopen this PR?

@wForget, I see. Thanks for your work!
@zhli1142015, could you take over this work if it is really necessary to you? You can create a new pr based on this one, with @wForget's work highlighted there.

**Comment by zhli1142015** [comment]
> > > @wForget, do you remember the landing gap of this pr? Thanks!
> > 
> > 
> > I remember the current status should be waiting for review, but I didn't move forward with it because I found `spark.gluten.sql.columnar.filescan=false` configuration. Do we need to reopen this PR?
> 
> @wForget, I see. Thanks for your work! @zhli1142015, could you take over this work if it is really necessary to you? You can create a new pr based on this one, with @wForget's work highlighted there.

Sure, thanks @wForget 's work.

**Pull request closed without merging**